### PR TITLE
enhance metrics on server and client

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

File: cmd/server/http/middleware/logging.go [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-	"github.com/gorilla/handlers"
-	"github.com/gorilla/mux"
 	"net/http"
 	"os"
+
+	"github.com/gorilla/handlers"
 )
 
-func Logging() mux.MiddlewareFunc {
+func Logging() func(http.Handler) http.Handler {
```
Comment (pimousse1099) [inline cmd/server/http/middleware/logging.go:11]: why getting rid of `MiddlewareFunc` ?
not to depend on `gorilla` ?

File: cmd/server/http/middleware/metrics.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/gol4ng/httpware/metrics"
+	"github.com/gol4ng/httpware/middleware"
+)
+
+func Metrics(recorder metrics.Recorder) func(http.Handler) http.Handler {
```
Comment (pimousse1099) [inline cmd/server/http/middleware/metrics.go:11]: explain what this middleware add 👍

File: cmd/server/http/middleware/logging.go (hunk 1 above)
Comment (pimousse1099) [reply cmd/server/http/middleware/logging.go:11]: explain what this middleware does 👍

File: cmd/server/http/middleware/prometheus.go [hunk 3]
```diff
@@ -1,18 +0,0 @@
-package middleware
-
-import (
-	"github.com/gorilla/mux"
-	"github.com/slok/go-http-metrics/metrics/prometheus"
-	"github.com/slok/go-http-metrics/middleware"
-	"net/http"
-)
-
-func Prometheus() mux.MiddlewareFunc {
```
Comment (pimousse1099) [inline cmd/server/http/middleware/prometheus.go:10]: why getting rid of this ?

File: cmd/server/http/server.go [hunk 4]
```diff
@@ -57,9 +57,9 @@ func Readiness(response http.ResponseWriter, _ *http.Request) {
 func getHttpHandler(container *service.Container) http.Handler {
 	r := mux.NewRouter()
 	r.Use(
-		middleware2.Logging(),
-		middleware2.Recovery(container.Cfg.Debug),
-		middleware2.Prometheus(),
+		middleware.Logging(),
```
Comment (pimousse1099) [inline cmd/server/http/server.go:60]: explain why this is done in this particular order 👍

File: go.mod [hunk 5]
```diff
@@ -2,6 +2,7 @@ module github.com/gol4ng/skeleton
 
 require (
 	github.com/fortytw2/leaktest v1.3.0 // indirect
+	github.com/gol4ng/httpware v0.0.0-20190724082439-5999a32c7adb
```
Comment (pimousse1099) [inline go.mod:5]: we might want to create a `v0.1.0` but it's ok for now

File: internal/metrics/tripperware.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"net/http"
+
+	"github.com/gol4ng/httpware"
+	"github.com/gol4ng/httpware/metrics"
+	"github.com/gol4ng/httpware/tripperware"
+)
+
+func Metrics(name string, recorder metrics.Recorder) httpware.Tripperware {
```
Comment (pimousse1099) [inline internal/metrics/tripperware.go]: maybe name it `Tripperware` instead of `Metrics` ?

Comment (pimousse1099) [reply internal/metrics/tripperware.go]: explain what this middleware does 👍

Comment (instabledesign) [reply internal/metrics/tripperware.go]: its not a middleware

File: cmd/server/http/middleware/prometheus.go (hunk 3 above)
Comment (instabledesign) [reply cmd/server/http/middleware/prometheus.go:10]: in general, we delete code when we move it in another location or because we don't need anymore

Comment (pimousse1099) [reply cmd/server/http/middleware/prometheus.go:10]: then explain where it has been moved or explain that it is not used anymore 👍

File: internal/metrics/tripperware.go (hunk 6 above)
Comment (pimousse1099) [reply internal/metrics/tripperware.go]: lol, well explain why this `not a middleware` is great and useful 😛

**Review by pimousse1099 — DISMISSED** [review]

File: cmd/server/http/middleware/prometheus.go (hunk 3 above)
Comment (instabledesign) [reply cmd/server/http/middleware/prometheus.go:10]: nop, im not going to explain all code modification

File: internal/metrics/tripperware.go (hunk 6 above)
Comment (instabledesign) [reply internal/metrics/tripperware.go]: it s not a middleware because it's a tripperware

**Pull request merged**